cpp/ql/src/semmle/code/cpp/models/implementations/Strcpy.qll

```diff
@@ -74,4 +75,23 @@ class StrcpyFunction extends ArrayFunction, DataFlowFunction, TaintFunction {
       output.isReturnValueDeref()
     )
   }
+
+  override predicate hasOnlySpecificReadSideEffects() { any() }
+
+  override predicate hasOnlySpecificWriteSideEffects() { any() }
+
+  predicate hasSpecificWriteSideEffect(ParameterIndex i, boolean buffer, boolean mustWrite) {
+    i = 0 and
+    buffer = true and
+    mustWrite = true
```

Comment:
What does `mustWrite` mean for a buffer? That the first byte is always overwritten? Is that also the case if the size argument to `strncpy` is 0?

Reply:
> What does `mustWrite` mean for a buffer? That the first byte is always overwritten?

That's what I had assumed.  It could be that the "whole buffer" is supposed to be overwritten, but in the case of `strcpy` it isn't clear what this "whole buffer" would mean.  Perhaps @dave-bartolomeo can clarify?

> Is that also the case if the size argument to strncpy is 0?

Good point.  I think it's something of an edge case (any any sane code would have tagged that data as empty in some way), but perhaps it's worth being ultra-cautious.

I'm now leaning towards changing this to `mustWrite = false`...

Reply:
If you set `mustWrite = false` (and run qlformat), then I can merge now without waiting for further input.

Reply:
Done.